shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Metadata.hs

```diff
@@ -0,0 +1,143 @@
+import Cardano.Ledger.ShelleyMA (MaryOrAllegra, ShelleyMAEra)
+import Cardano.Ledger.ShelleyMA.Scripts ()
+import Control.DeepSeq (deepseq)
+import Data.Coders
+import Data.Map.Strict (Map)
+import Data.MemoBytes
+import Data.Sequence.Strict (StrictSeq)
+import Data.Typeable (Typeable)
+import Data.Word (Word64)
+import GHC.Generics (Generic)
+import NoThunks.Class
+import Shelley.Spec.Ledger.MetaData
+  ( MetaDataHash (..),
+    MetaDatum,
+    ValidateMetadata (..),
+    validMetaDatum,
+  )
+
+-- | Raw, un-memoised metadata type
+data MetadataRaw era = MetadataRaw
+  { mdScriptPreimages :: !(StrictSeq (Core.Script era)),
+    -- | Unstructured metadata "blob"
+    mdBlob :: !(Map Word64 MetaDatum)
+  }
+  deriving (Generic)
+
+deriving instance (Core.ChainData (Core.Script era)) => Eq (MetadataRaw era)
+
+deriving instance (Core.ChainData (Core.Script era)) => Show (MetadataRaw era)
+
+deriving instance
+  (Core.ChainData (Core.Script era)) =>
+  NoThunks (MetadataRaw era)
+
+newtype Metadata era = MetadataWithBytes (MemoBytes (MetadataRaw era))
+  deriving (Typeable)
+  deriving newtype (ToCBOR)
+
+deriving newtype instance
+  (Era era, Core.ChainData (Core.Script era)) =>
+  Eq (Metadata era)
+
+deriving newtype instance
+  (Era era, Core.ChainData (Core.Script era)) =>
+  Show (Metadata era)
+
+deriving newtype instance
+  (Era era, Core.ChainData (Core.Script era)) =>
+  NoThunks (Metadata era)
+
+pattern Metadata ::
+  ( Core.AnnotatedData (Core.Script era),
+    Ord (Core.Script era)
+  ) =>
+  StrictSeq (Core.Script era) ->
+  Map Word64 MetaDatum ->
+  Metadata era
+pattern Metadata sp blob <-
+  MetadataWithBytes (Memo (MetadataRaw sp blob) _)
+  where
+    Metadata sp blob =
+      MetadataWithBytes $
+        memoBytes
+          (encMetadataRaw $ MetadataRaw sp blob)
+
+{-# COMPLETE Metadata #-}
+
+type instance
+  Core.Metadata (ShelleyMAEra (ma :: MaryOrAllegra) c) =
+    Metadata (ShelleyMAEra (ma :: MaryOrAllegra) c)
+
+instance
+  ( Crypto c,
+    Typeable ma,
+    Core.AnnotatedData (Core.Script (ShelleyMAEra ma c))
+  ) =>
+  ValidateMetadata (ShelleyMAEra (ma :: MaryOrAllegra) c)
+  where
+  hashMetadata = MetaDataHash . hashWithSerialiser toCBOR
+
+  validateMetadata (Metadata sp blob) = deepseq sp $ all validMetaDatum blob
+
+--------------------------------------------------------------------------------
+-- Serialisation
+--------------------------------------------------------------------------------
+
+-- | Encode Metadata
+encMetadataRaw ::
+  (Core.AnnotatedData (Core.Script era)) =>
+  MetadataRaw era ->
+  Encode ( 'Closed 'Dense) (MetadataRaw era)
+encMetadataRaw (MetadataRaw sp blob) =
+  Rec MetadataRaw
+    !> E encodeFoldable sp
+    !> To blob
+
+instance
+  (Era era, Core.AnnotatedData (Core.Script era)) =>
+  FromCBOR (Annotator (MetadataRaw era))
+  where
+  fromCBOR =
+    decode
+      ( Ann (RecD MetadataRaw)
+          <*! D (sequence <$> decodeStrictSeq fromCBOR)
+          <*! Ann From
+      )
```

Comment:
I thought that we wanted the deserialization to be backwards compatible.

Reply:
Yes, I am an idiot and totally forgot about this. Will fix.

Reply:
you are anything but that! and thanks again for taking this one on!

Reply:
@redxaxder and I propose this: https://github.com/input-output-hk/cardano-ledger-specs/pull/1994/files#diff-03ecce724e5f30f7e4d885d88b81708a4bc1bc0451d2aa8b82137f7791f5c07bR214-R219